### PR TITLE
close _eventStream when stopping watcher

### LINKED ISSUE #40888: Disposing the FileSystemWatcher does not close associated file descriptors
Filed this against Mono, which uses corefx, but the problem also affects corefx.

https://github.com/mono/mono/issues/16709

## Steps to Reproduce

Related problem - https://github.com/mono/mono/issues/15931

1. Sample project - https://github.com/mrward/test-file-watcher-dispose - multi-target project that builds .NET 4.7.2 and .NET Core 2.1
2. Build this project.
3. From the command line run `dotnet bin/debug/netcoreapp2.1/FileWatcherDisposeTest.dll`
4. Run `lsof -p pid`. Process id is output from the console app.
5. See that the file watcher creates file descriptors for several directories.
6. After 10 seconds the file watcher will be disposed.
7. Run `lsof -p pid` again.

### Current Behavior

lsof still lists the directories.

```
dotnet 3972 user    4r DIR   1,12      170   51740695 /Volumes/Drive/projects/tests/FileWatcherDisposeTest/FileWatcherDisposeTest/bin/Debug/net472
dotnet 3972 user    5r DIR   1,12      136   51740600 /Volumes/Drive/projects/tests/FileWatcherDisposeTest/FileWatcherDisposeTest/bin/Debug
dotnet 3972 user    6r DIR   1,12      102   51740599 /Volumes/Drive/projects/tests/FileWatcherDisposeTest/FileWatcherDisposeTest/bin
dotnet 3972 user    7r DIR   1,12      204   51740594 /Volumes/Drive/projects/tests/FileWatcherDisposeTest/FileWatcherDisposeTest
dotnet 3972 user    8r DIR   1,12      170   51740593 /Volumes/Drive/projects/tests/FileWatcherDisposeTest
dotnet 3972 user    9r DIR   1,12     6324     185300 /Volumes/Drive/projects/tests
dotnet 3972 user   10r DIR   1,12     2312        278 /Volumes/Drive/projects
dotnet 3972 user   11r DIR   1,12      646          2 /Volumes/Drive
dotnet 3972 user   12r DIR    1,4      160     420284 /Volumes
```
If you change the console project to run GC.Collect after the dispose does not fix the problem (this does work with Mono)

### Expected Behavior

File watcher directories not listed by lsof.

Note that 

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview9-014004
 Commit:    8e7ef240a5

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.14
 OS Platform: Darwin
 RID:         osx.10.14-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.0.100-preview9-014004/

Host (useful for support):
  Version: 3.0.0-preview9-19423-09
  Commit:  2be172345a

.NET Core SDKs installed:
  2.1.701 [/usr/local/share/dotnet/sdk]
  3.0.100-preview9-014004 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.12 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.12 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview9.19424.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.12 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview9-19423-09 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

### PR DESCRIPTION
**Pull request by wfurt**
When even stream is created it needs to be properly cleaned up in order to close associated file handles and associated resources. 
https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html

We need to call  FSEventStreamInvalidate and FSEventStreamRelease. Conveniently, it seems like this is all encapsulated in SafeEventStreamHandle's ReleaseHandle. 

I run repro code from #40888 in loop and I did not see any descriptor leak. (The leak was pretty obvious without the fix) 
I may spent more time tomorrow thinking if there is some good way how to add test but it may be tricky to verify OS resources. 

fixes #40888

### CODE REVIEW

**Comment by Therzok** [comment]
You can probably use [CFGetRetainCount](https://developer.apple.com/documentation/corefoundation/1521288-cfgetretaincount?language=objc) before releasing the last ref to validate it's 1.

**Comment by Therzok** [comment]
Can confirm, mono side is fixed too 🎉

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 1]
```diff
@@ -271,6 +271,8 @@ private void CancellationCallback()
                     finally
                     {
                         StaticWatcherRunLoopManager.UnscheduleFromRunLoop(_eventStream);
+                        _eventStream.Close();
+                        _eventStream = null;
```
Comment (stephentoub) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: When can this be invoked?  Is it possible we could be nulling out _eventStream and have a race condition with someone else trying to dereference it?

Comment (wfurt) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: I did another check and I don't think so. we use _eventStream to start eventloop but it is not passed or used in callback with events.
CancellationCallback() has check for null and it calls FSEventStreamSto() first to stop loop and than we take steps to invalidate and release according to Apple's documentation.  
Only problem I could see is that the neither `_stopping` nor check for `_eventStream != null` is atomic.  Also `_disposed` is not set and check in atomic way in `FileSystemWatcher.cs`. 
I think there could be race if Dispose() is called concurrently from multiple threads. 
If you think it is worth it, I can assign `_eventStream` to local variable first and then check for null and call Close().

**Comment by wfurt** [comment]
test failures in System.Security.Cryptography are unrelated. Please let me know if there is anything outstanding @JeremyKuhne and @stephentoub.
I would like to get this into master and work on 2.x port.

**Comment by danmosemsft** [comment]
@carlossanlop perhaps you can review?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs [hunk 2]
```diff
@@ -72,6 +72,7 @@ private void StopRaisingEvents()
             {
                 _cancellation = null;
                 token.Cancel();
+                token.Dispose();
```
Comment (carlossanlop) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: @wfurt I have a question: why do we *only* cancel and dispose the `token` here? The RunningInstance class receives a copy of this token, so why not perform these two actions inside `CancellationCallback` too? (follow up of Marius' question).

**Comment by stephentoub** [comment]
@wfurt, what's the status of this PR?  Can it be made ready to merge, or closed?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs (hunk 2 above)
Comment (wfurt) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs]: CancellationCallback() did not do anything with the token. It is StartRaisingEvents() where token was created and symmetrically StopRaisingEvents() where token is destroyed. Existing `_cancellation = null` should be sufficient IMHO to dispose token. I only add explicit Dispose() call to make it explicit and not wait for GC. I did not want to introduce different life cycle for the token and worry about what changed.

**Comment by wfurt** [comment]
Tests are failing in System.Security.Cryptography on OSX and x86 Windows. That seems unrelated to this change. 

As far as the test I could not come up with one I would like. Best I think I can do is to create and destroy watcher in loop and make loop count bigger than default handle limit. But CI does manipulate limits AFAIK and well as that may differ across systems. Also tests  trying to exhaust system resources are potentially dangerous. Let me know @carlossanlop if you have suggestions.

**Comment by carlossanlop** [comment]
> _As far as the test I could not come up with one I would like. Best I think I can do is to create and destroy watcher in loop and make loop count bigger than default handle limit. But CI does manipulate limits AFAIK and well as that may differ across systems. Also tests trying to exhaust system resources are potentially dangerous. Let me know @carlossanlop if you have suggestions._

I noticed the original issue [#40888](https://github.com/dotnet/corefx/issues/40888) has a small file watcher test here: https://github.com/mrward/test-file-watcher-dispose/blob/master/Program.cs

Would it make sense to write a unit test that looks somewhat like in that code, or would we encounter the CI limit you mentioned, @wfurt ?

**Comment by Therzok** [comment]
I know this is icky, but what if we had something like:
```csharp
var streamHandle = // use reflection to get the native handle
var ref = new WeakReference (streamHandle);
...
// do the test

Assert.AreEqual (1, CFRetainCount (streamHandle.Handle));

// dispose the watcher and wait for GC to run

Assert.IsNull (ref.Target);
```

**Comment by wfurt** [comment]
I added test and verified it fails in CI without the fix same way as #40888
https://helix.dot.net/api/2019-06-17/jobs/7d88982f-e479-45e6-b82f-36c65c3e89bd/workitems/System.IO.FileSystem.Watcher.Tests/console

```
   System.IO.Tests.FileSystemWatcherTests_Unix.Watcher_Usage_DoesNotLeak [STARTING]
Limits are 10240 and 18446744073709551615
File open failed at 10067 with Too many open files
    System.IO.Tests.FileSystemWatcherTests_Unix.Watcher_Usage_DoesNotLeak [FINISHED] Time: 0.6855082s
```

I use new trick to push test to the and making sure that it does not run with any other tests. 
That allows to use XUnit output helper . (which is not available for "remote" exec in separate process.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+using Microsoft.DotNet.XUnitExtensions;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.IO.Tests
+{
+    public partial class DangerousFileSystemWatcherTests
+    {
+        [ConditionalFact]
+        [OuterLoop("Slow test with significant resource ussage.")]
```
Comment (carlossanlop) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Unix.cs]:
```suggestion
        [OuterLoop("Slow test with significant resource usage.")]
```

**Comment by wfurt** [comment]
System.Runtime tests failures are unrelated. They also fail in master. The Musl run did not producer any reasonable output., That seems like infrastructure problem.

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

**Pull request merged**